File: its-my-training-day/algorithms/leetcode-top-75/questions/graphs/union-find/connected-components/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def number_of_connected_components(n: int, edges: List[List[int]]) -> int:
    parent = [i for i in range(n)]
    count = n

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    i = 0
    while i < len(edges):
        a = edges[i][0]
        b = edges[i][1]
        pa = find(a)
        pb = find(b)
        if pa != pb:
            parent[pb] = pa
            count -= 1
        i += 1
    return count

File: its-my-training-day/algorithms/leetcode-top-75/questions/graphs/union-find/connected-components/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

int numberOfConnectedComponents(int n, const std::vector<std::vector<int>>& edges) {
    std::vector<int> parent(n);
    for (int i = 0; i < n; ++i) parent[i] = i;
    int count = n;
    auto find = [&](int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (parent[x] != x) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    };
    for (int i = 0; i < static_cast<int>(edges.size()); ++i) {
        int a = edges[i][0];
        int b = edges[i][1];
        int pa = find(a);
        int pb = find(b);
        if (pa != pb) {
            parent[pb] = pa;
            count -= 1;
        }
    }
    return count;
}
